CASSANDRA-17186: Guardrail for number of partition keys on IN queries

tejavadali opened the pull request:
See [CASSANDRA-17186](https://issues.apache.org/jira/browse/CASSANDRA-17186)

patch by tejavadali; reviewed by Andres de la Peña, Paulo Motta for CASSANDRA-17186

adelapena commented on src/java/org/apache/cassandra/cql3/statements/SelectStatement.java:
```diff
@@ -527,6 +528,13 @@ private ReadQuery getSliceCommands(QueryOptions options, ColumnFilter columnFilt
+        if (restrictions.keyIsInRelation())
+        {
+            Guardrails.partitionKeysInSelect.guard(keys.size(),
+                                                   String.format("partition keys in IN clause on table %s", table.name),
```
We can put the `partition keys in IN clause on table` message in the message provider at `Guardrails.partitionKeysInSelect`, so here we only need to pass the table name.

tejavadali replied:
👍

adelapena commented on src/java/org/apache/cassandra/cql3/statements/SelectStatement.java:
```diff
@@ -527,6 +528,13 @@ private ReadQuery getSliceCommands(QueryOptions options, ColumnFilter columnFilt
+        {
+            Guardrails.partitionKeysInSelect.guard(keys.size(),
+                                                   String.format("partition keys in IN clause on table %s", table.name),
+                                                   null);
```
Here we need to pass the `ClientState` for the query, so the guardrail is not applied to super user and internal queries. 

This would require to modify the signature of this method and also the signature of `SelectStatement#getQuery(QueryOptions, ColumnFilter, int, int)`. I just did that as part of the [PR](https://github.com/apache/cassandra/pull/1444/files) for [CASSANDRA-17187](https://issues.apache.org/jira/browse/CASSANDRA-17187), you can take the relevant changes on signatures from that patch.

Also, we would need a test to verify that the guardrail is skipped for those users. There are multiple tests named `testExcludedUsers` around that are used for equivalent checks for other guardrails.

tejavadali replied:
👍

adelapena commented on test/unit/org/apache/cassandra/db/guardrails/GuardrailPartitionKeysInSelectTest.java:
```diff
@@ -0,0 +1,79 @@
+    @Before
+    public void setupTest()
+    {
+        createKeyspace("CREATE KEYSPACE IF NOT EXISTS partition_keys_in_select_test_ks " +
+                       "with replication = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 } " +
+                       "and durable_writes = false");
+        createTable("CREATE TABLE IF NOT EXISTS partition_keys_in_select_test_ks.partition_keys_in_select_test " +
+                    "(k INT, c INT, v TEXT, PRIMARY KEY(k, c))");
+    }
+
+    @Test
+    public void testSelectStatementAgainstInClausePartitionKeys() throws Throwable
+    {
+        assertValid("SELECT k, c, v FROM partition_keys_in_select_test_ks.partition_keys_in_select_test " +
+                    "WHERE k=10");
+
+        assertValid("SELECT k, c, v FROM partition_keys_in_select_test_ks.partition_keys_in_select_test " +
+                    "WHERE k IN (2, 3)");
+
+        assertValid("SELECT k, c, v FROM partition_keys_in_select_test_ks.partition_keys_in_select_test " +
+                    "WHERE k = 2 and c IN (2, 3, 4, 5, 6, 7)");
+
+        assertWarns("Query with partition keys in IN clause on table partition_keys_in_select_test, " +
+                    "with number of partition keys 4 exceeds warning threshold of 3.",
+                    "SELECT k, c, v FROM partition_keys_in_select_test_ks.partition_keys_in_select_test " +
+                        "WHERE k IN (2, 3, 4, 5)");
+
+        assertAborts("Aborting query with partition keys in IN clause on table partition_keys_in_select_test, " +
+                     "number of partition keys 6 exceeds abort threshold of 5." ,
+                     "SELECT k, c, v FROM partition_keys_in_select_test_ks.partition_keys_in_select_test " +
+                     "WHERE k IN (2, 3, 4, 5, 6, 7)"
+                     );
```
Instead of creating a keyspace we can use the default keyspace for tests and autogenerated table name, similarly to what's done for example in [`GuardrailPageSizeTest`](https://github.com/apache/cassandra/blob/trunk/test/unit/org/apache/cassandra/db/guardrails/GuardrailPageSizeTest.java). It's not a big difference but I think it makes the test a bit more concise and, more importantly, it makes sure that the table is removed after finishing the test:
```suggestion
        createTable("CREATE TABLE %s (k INT, c INT, v TEXT, PRIMARY KEY(k, c))");
    }

    @Test
    public void testSelectStatementAgainstInClausePartitionKeys() throws Throwable
    {
        assertValid("SELECT k, c, v FROM %s WHERE k=10");

        assertValid("SELECT k, c, v FROM %s WHERE k IN (2, 3)");

        assertValid("SELECT k, c, v FROM %s WHERE k = 2 and c IN (2, 3, 4, 5, 6, 7)");

        assertWarns(format("Query with partition keys in IN clause on table %s, " +
                           "with number of partition keys 4 exceeds warning threshold of 3.",
                           currentTable()),
                    "SELECT k, c, v FROM %s WHERE k IN (2, 3, 4, 5)");

        assertAborts(format("Aborting query with partition keys in IN clause on table %s, " +
                            "number of partition keys 6 exceeds abort threshold of 5.",
                            currentTable()),
                     "SELECT k, c, v FROM %s WHERE k IN (2, 3, 4, 5, 6, 7)");
```

tejavadali replied:
👍

tejavadali commented:
@adelapena Addressed your comments, the PR is ready for another look. Ill resolve the conflicts at the end because as I resolve them, they keep coming.

adelapena commented on src/java/org/apache/cassandra/cql3/statements/SelectStatement.java:
```diff
@@ -247,7 +248,8 @@ public void validate(ClientState state) throws InvalidRequestException
 
         Selectors selectors = selection.newSelectors(options);
-        ReadQuery query = getQuery(options, selectors.getColumnFilter(), nowInSec, userLimit, userPerPartitionLimit, pageSize);
+        ReadQuery query = getQuery(options, ClientState.forInternalCalls(), selectors.getColumnFilter(),
```
Here we should use the `ClientState` that is contained in the `QueryState` argument, so it would be:
```suggestion
        ReadQuery query = getQuery(options, state.getClientState(), selectors.getColumnFilter(),
```

tejavadali replied:
👍

adelapena commented on test/unit/org/apache/cassandra/db/guardrails/GuardrailPartitionKeysInSelectTest.java:
```diff
@@ -0,0 +1,80 @@
+    @Test
+    public void testExcludedUsers() throws Throwable
+    {
+        testExcludedUsers(() -> format("SELECT k, c, v FROM %s WHERE k IN (2, 3, 4, 5)", tableName),
+                          () -> format("SELECT k, c, v FROM %s WHERE k IN (2, 3, 4, 5, 6, 7)", tableName));
```
We don't need to format the queries with the table name:
```suggestion
        testExcludedUsers(() -> "SELECT k, c, v FROM %s WHERE k IN (2, 3, 4, 5)",
                          () -> "SELECT k, c, v FROM %s WHERE k IN (2, 3, 4, 5, 6, 7)");
```

tejavadali replied:
👍

tejavadali commented:
@adelapena added your suggestions. Thanks :)

adelapena commented:
Great, thanks, the PR looks good to me. I'll run CI tests once it's rebased on trunk. The relevant changes introduced in [CASSANDRA-17353](https://issues.apache.org/jira/browse/CASSANDRA-17353) are just the unnested config and renaming "abort" thresholds by "fail".

Commit:
CASSANDRA-17186: Guardrail for number of partition keys on IN queries

patch by tejavadali; reviewed by Andres de la Peña, Paulo Motta for CASSANDRA-17186

Commit:
Address review comments

Commit:
Fix minor review

tejavadali commented:
Resolved conflits.

adelapena commented on conf/cassandra.yaml:
```diff
@@ -1616,3 +1616,8 @@ drop_compact_storage_enabled: false
 # Guardrail to allow/disallow list operations that require read before write, i.e. setting list element by index and
 # removing list elements by either index or value. Defaults to true.
 # read_before_write_list_operations_enabled: true
+# Guardrail to warn or abort when a client issued SELECT statement has more partition keys in the IN clause than threshold.
+# The two thresholds default to -1 to disable.
+#     partition_keys_in_select:
+#         warn_threshold: -1
+#         abort_threshold: -1
```
This should be updated to the new flat format:
```suggestion
# Guardrail to warn or fail when querying with an IN restriction selecting
# more partition keys than threshold.
# The two thresholds default to -1 to disable. 
partition_keys_in_select_warn_threshold: -1
partition_keys_in_select_fail_threshold: -1
```

adelapena commented on src/java/org/apache/cassandra/db/guardrails/Guardrails.java:
```diff
@@ -127,6 +127,20 @@
+                                     "partition keys %s exceeds warning threshold of %s.",
+                                     what, value, threshold)
+                            : format("Aborting query with partition keys in IN clause on table %s, " +
+                                     "number of partition keys %s exceeds abort threshold of %s.",
```
```suggestion
                                     "number of partition keys %s exceeds fail threshold of %s.",
```

adelapena commented on src/java/org/apache/cassandra/db/guardrails/Guardrails.java:
```diff
@@ -351,6 +365,24 @@ public void setReadBeforeWriteListOperationsEnabled(boolean enabled)
     }
 
+    @Override
+    public void setPartitionKeysInSelectThreshold(int warn, int abort)
+    {
+        DEFAULT_CONFIG.setPartitionKeysInSelectThreshold(warn, abort);
```
```suggestion
    public void setPartitionKeysInSelectThreshold(int warn, int fail)
    {
        DEFAULT_CONFIG.setPartitionKeysInSelectThreshold(warn, fail);
```

adelapena commented on test/unit/org/apache/cassandra/db/guardrails/GuardrailPartitionKeysInSelectTest.java:
```diff
@@ -0,0 +1,76 @@
+public class GuardrailPartitionKeysInSelectTest extends ThresholdTester
+{
+    private static final int PARTITION_KEYS_SELECT_WARN_THRESHOLD = 3;
+    private static final int PARTITION_KEYS_SELECT_ABORT_THRESHOLD = 5;
```
```suggestion
    private static final int PARTITION_KEYS_SELECT_FAIL_THRESHOLD = 5;
```

adelapena commented on test/unit/org/apache/cassandra/db/guardrails/GuardrailPartitionKeysInSelectTest.java:
```diff
@@ -0,0 +1,76 @@
+                    "SELECT k, c, v FROM %s WHERE k IN (2, 3, 4, 5)");
+
+        assertFails(String.format("Aborting query with partition keys in IN clause on table %s, " +
+                     "number of partition keys 6 exceeds abort threshold of 5.", tableName) ,
```
```suggestion
                     "number of partition keys 6 exceeds fail threshold of 5.", tableName) ,
```

adelapena reviewed (commented):
Looks good, it seems that the config on cassandra.yaml has been missed on the rebase, and there are some uses of "abort" instead of "fail".

tejavadali committed:
Update conf/cassandra.yaml

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

tejavadali committed:
Update src/java/org/apache/cassandra/db/guardrails/Guardrails.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

tejavadali committed:
Update test/unit/org/apache/cassandra/db/guardrails/GuardrailPartitionKeysInSelectTest.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

tejavadali committed:
Update test/unit/org/apache/cassandra/db/guardrails/GuardrailPartitionKeysInSelectTest.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

tejavadali committed:
Update src/java/org/apache/cassandra/db/guardrails/Guardrails.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

tejavadali commented:
Sorry about that, I missed those during the merge.

tejavadali commented:
@adelapena please let me know if you need anything else on this from my side. Else, you can submit this for CI.

adelapena commented:
I have already run CI for this. The results are on the JIRA ticket, and they look good to me. Now we need the approval of a second reviewer before committing.

bereng commented on src/java/org/apache/cassandra/config/GuardrailsOptions.java:
```diff
@@ -56,6 +56,7 @@
     
     private final Config config;
 
+
```
Needed?

bereng commented on src/java/org/apache/cassandra/config/GuardrailsOptions.java:
```diff
@@ -194,6 +197,31 @@ public int getMaterializedViewsPerTableWarnThreshold()
         return config.materialized_views_per_table_warn_threshold;
     }
 
+    @Override
```
Here you have 2 getters but only one setter for both. This is usually a pain bc it forces you to do a read before write when you want to set only one. It's a nit but I thought I'd raise it.

adelapena replied:
This single setter approach is used by all thresholds. It was discussed during [CASSANDRA-17147](https://github.com/apache/cassandra/pull/1321#), see [this comment](https://github.com/apache/cassandra/pull/1321#discussion_r760697524). The idea is that the validation of each warn/fail thresholds is tied to the other, so in the end if you want to change one you nevertheless need to know the other.

bereng commented on conf/cassandra.yaml:
```diff
@@ -1616,3 +1616,8 @@ drop_compact_storage_enabled: false
+# Guardrail to warn or fail when querying with an IN restriction selecting
+# more partition keys than threshold.
+# The two thresholds default to -1 to disable. 
+partition_keys_in_select_warn_threshold: -1
```
Why not leave these commented out like the others?

adelapena replied:
Good catch, this should be commented. I've missed this one.

tejavadali replied:
👍

adelapena commented on test/unit/org/apache/cassandra/db/guardrails/GuardrailPartitionKeysInSelectTest.java:
```diff
@@ -0,0 +1,76 @@
+    public GuardrailPartitionKeysInSelectTest()
+    {
+        super(PARTITION_KEYS_SELECT_WARN_THRESHOLD,
+              PARTITION_KEYS_SELECT_ABORT_THRESHOLD,
```
```suggestion
              PARTITION_KEYS_SELECT_FAIL_THRESHOLD,
```

tejavadali committed:
Update test/unit/org/apache/cassandra/db/guardrails/GuardrailPartitionKeysInSelectTest.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

Commit:
Fix review comments

tejavadali commented:
Resolved review comments, ready for another look :)

tejavadali committed:
Merge branch 'trunk' into CASSANDRA-17186
